DICOM/src/Microsoft.Health.DeID.SharedLib/RedactFunction.cs

```diff
@@ -0,0 +1,147 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Linq;
+using System.Text.RegularExpressions;
+using EnsureThat;
+using Microsoft.Health.Dicom.DeID.SharedLib.Models;
+using Microsoft.Health.Dicom.DeID.SharedLib.Settings;
+
+namespace Microsoft.Health.Dicom.DeID.SharedLib
+{
+    public class RedactFunction
+    {
+        private static readonly int AgeThreshold = 89;
+        private static readonly string ReplacementDigit = "0";
+        private static readonly int InitialDigitsCount = 3;
+        private readonly RedactSetting _redactSetting;
+
+        public RedactFunction(RedactSetting redactSetting = null)
+        {
+            _redactSetting = redactSetting ?? new RedactSetting();
+        }
+
+        public string RedactDate(string dateString, string inputDateFormat = null, string outputDateFormat = null, IFormatProvider provider = null)
+        {
+            if (string.IsNullOrEmpty(dateString))
+            {
+                return null;
+            }
+
+            if (_redactSetting.EnablePartialDatesForRedact)
+            {
+                DateTimeOffset date = DateTimeUtility.ParseDateTimeString(dateString, inputDateFormat, provider, DeIDGlobalSettings.DateFormat);
+                return DateTimeUtility.IndicateAgeOverThreshold(date)
+                    ? null
+                    : new DateTimeOffset(date.Year, 1, 1, 0, 0, 0, new TimeSpan(0, 0, 0)).ToString(outputDateFormat ?? DeIDGlobalSettings.DateFormat);
+            }
+
+            return null;
+        }
+
+        public string RedactDateTime(string dateTimeString, string inputDateTimeFormat = null, string outputDateFormat = null, IFormatProvider provider = null)
+        {
+            if (string.IsNullOrEmpty(dateTimeString))
+            {
+                return null;
+            }
+
+            if (_redactSetting.EnablePartialDatesForRedact)
+            {
+                DateTimeOffset date = DateTimeUtility.ParseDateTimeString(dateTimeString, inputDateTimeFormat, provider);
+                return DateTimeUtility.IndicateAgeOverThreshold(date)
+                    ? null
+                    : new DateTimeOffset(date.Year, 1, 1, 0, 0, 0, new TimeSpan(0, 0, 0)).ToString(outputDateFormat ?? DeIDGlobalSettings.DateTimeFormat);
+            }
+
+            return null;
+        }
+
+        public DateTimeOffset? RedactDateTime(DateTimeOffset dateTime)
+        {
+            EnsureArg.IsNotNull<DateTimeOffset>(dateTime, nameof(dateTime));
+
+            if (_redactSetting.EnablePartialDatesForRedact)
```

Comment:
🟢 [Discussion]
I personally prefer below style:
```
if (_redactSetting.EnablePartialDatesForRedact && !DateTimeUtility.IndicateAgeOverThreshold(dateTime))
{
    return new DateTimeOffset(dateTime.Year, 1, 1, 0, 0, 0, default);
}

return null;
```

But if the logic becomes more complex, nested if/else might be inevitable.